cclib/method/bader.py

```diff
@@ -0,0 +1,219 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2020, the cclib development team
+#
+# This file is part of cclib (http://cclib.github.io) and is distributed under
+# the terms of the BSD 3-Clause License.
+
+"""Calculation of Bader's QTAIM charges based on data parsed by cclib."""
+import random
+import numpy
+import logging
+import math
+
+from cclib.method.calculationmethod import Method
+from cclib.method.volume import electrondensity_spin
+from cclib.parser.utils import convertor
+
+# Distance between two adjacent grids (sqrt[2] or sqrt[3] for uniform Cartesian grid)
+griddist = [
+    [[1.73205, 1.41421, 1.73205], [1.41421, 1, 1.41421], [1.73205, 1.41421, 1.73205],],
+    [[1.41421, 1, 1.41421], [1, 1, 1], [1.41421, 1, 1.41421]],
+    [[1.73205, 1.41421, 1.73205], [1.41421, 1, 1.41421], [1.73205, 1.41421, 1.73205],],
+]
+
+
+class MissingInputError(Exception):
+    pass
+
+
+class Bader(Method):
+    """Bader's QTAIM charges."""
+
+    # All of these are required for QTAIM charges.
+    required_attrs = ("homos", "mocoeffs", "nbasis", "gbasis")
+
+    def __init__(self, data, volume, progress=None, loglevel=logging.INFO, logname="Log"):
+        super(Bader, self).__init__(data, progress, loglevel, logname)
+
+        self.volume = volume
+        self.fragresults = None
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "Bader's QTAIM charges of {}".format(self.data)
+
+    def __repr__(self):
+        """Return a representation of the object."""
+        return "Bader({})".format(self.data)
+
+    def _check_required_attributes(self):
+        super(Bader, self)._check_required_attributes()
+
+    def calculate(self, indices=None, fupdate=0.05):
+        """Calculate Bader's QTAIM charges using on-grid algorithm proposed by Henkelman group
+           in doi:10.1016/j.commatsci.2005.04.010
+           
+           Cartesian, uniformly spaced grids are assumed for this function.
+           """
+
+        def cartesian_dist(pt1, pt2):
```

Comment:
I would make this a private function at the module level.

Reply:
I've made this function private. Thanks!